src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java

```
@@ -0,0 +1,189 @@
+        try {
+            SecurityUtils.setAuthorizationObject();
+
+            Set<LearningGoal> learningGoals;
+            if (learningObject instanceof Exercise exercise) {
+                learningGoals = exerciseRepository.findByIdWithLearningGoals(exercise.getId()).map(Exercise::getLearningGoals).orElse(null);
+            }
+            else if (learningObject instanceof LectureUnit lectureUnit) {
+                learningGoals = lectureUnitRepository.findByIdWithLearningGoals(lectureUnit.getId()).map(LectureUnit::getLearningGoals).orElse(null);
+            }
+            else {
+                return;
+            }
+
+            if (learningGoals == null) {
+                // Learning goals couldn't be loaded, the exercise/lecture unit might have already been deleted
+                return;
+            }
+
+            // The participant is normally a single user, but we need to support teams as well
+            participant.getParticipants().forEach(user -> {
+                learningGoals.forEach(learningGoal -> {
+                    updateProgress(learningGoal.getId(), user);
+                });
+            });
+        }
+        catch (Exception e) {
+            logger.error("Exception while updating progress for learning goal", e);
+        }
+    }
+
+    /**
+     * Updates the progress value (and confidence score) of the given learning goal and user
+     * @param learningGoalId The id of the learning goal to update the progress for
+     * @param user The user for which the progress should be updated
+     */
+    private void updateProgress(Long learningGoalId, User user) {
+        var learningGoal = learningGoalRepository.findByIdWithExercisesAndLectureUnitsAndCompletions(learningGoalId).orElse(null);
+
+        if (user == null || learningGoal == null) {
+            // If the user or learning goal no longer exist, there is nothing to do
+            return;
```

Would be interesting for debugging purposes to log a message here instead of completely silently returning?